Repo tests - conversion to pytest

bigtedde opened the pull request:
The current repository tests are written with the unittest framework. I have began converting these to pytest, starting with just a few tests so as to make sure I am on the right track with these changes.

Very rough draft, started in a separate test file for now. If I am on the right track I will merge changes into original test file and continue converting.

bigtedde commented on src/borg/testsuite/repo_new.py:
```diff
@@ -0,0 +1,257 @@
+
+class TestRepository(RepositoryTestCaseBase):
+    @pytest.fixture()
+    def repository(self, tmpdir):
```
Repo fixture to replace the repo setup and teardown.

bigtedde replied:
We did not have this yet, so I added it as a fixture to conftest

bigtedde commented on src/borg/testsuite/repo_new.py:
```diff
@@ -0,0 +1,257 @@
+            yield repository
+
+    def test1(self, repository):
+        for x in range(100):
```
Guess we should rename these tests to something more descriptive?

ThomasWaldmann replied:
sure.

ThomasWaldmann commented on src/borg/testsuite/repo_new.py:
```diff
@@ -0,0 +1,257 @@
+
+
+class RepositoryTestCaseBase:
+    def reopen(self, repository, exclusive=UNSPECIFIED):
+        repository.__exit__(None, None, None)
+        repository.__enter__()
+        repository.exclusive = exclusive
```
this could maybe become a `Repository` method.

ThomasWaldmann replied:
btw, one reason why this was done like that is that the unittest-style base class had a setup/teardown method that entered/exited the repository context. So the exit/enter in `reopen` was to work within that given setup/teardown environment.

bigtedde replied:
With this in mind I tried to replicate that behavior in my pytest fixtures for `repository` and `reopen_repository`. I just pushed these fixtures - is this what you had in mind?

ThomasWaldmann commented on src/borg/testsuite/repo_new.py:
```diff
@@ -0,0 +1,257 @@
+        repository.__enter__()
+        repository.exclusive = exclusive
+
+    def add_keys(self, repository):
+        repository.put(H(0), fchunk(b"foo"))
+        repository.put(H(1), fchunk(b"bar"))
+        repository.put(H(3), fchunk(b"bar"))
+        repository.commit(compact=False)
+        repository.put(H(1), fchunk(b"bar2"))
+        repository.put(H(2), fchunk(b"boo"))
+        repository.delete(H(3))
```
i don't see a call of this method.

bigtedde replied:
It gets called by some of the later tests in this file

ThomasWaldmann reviewed (changes requested):
Can you delete the old tests which these tests replace?

ThomasWaldmann commented:
Please rebase on current master.

bigtedde commented:
> Can you delete the old tests which these tests replace?

Eventually I plan to replace the whole file. Since I have only replaced a portion of the tests so far, I believe deleting some of the helper functions and tests would cause others not to work properly.

bigtedde commented on src/borg/repository.py:
```diff
@@ -463,6 +463,10 @@ def open(self, path, exclusive, lock_wait=None, lock=True):
         self.id = unhexlify(self.config.get("repository", "id").strip())
         self.io = LoggedIO(self.path, self.max_segment_size, self.segments_per_dir)
 
+    def reopen(self, exclusive=True):
```
The original test suite has a reopen method that created a new repository object between the `__exit__` and `__enter__` 

I didn't think it would be correct to do the same thing here, although I was not sure how to mimc this behavior, and thus all the tests that call this reopen function are having issues. I spent  _several_  hours trying get reopen to work as it does in the original test file, to no success.

RonnyPfannschmidt replied:
Instead of reopen, have a fixture that's returns a non-reentreant context manager which yields the repo on entering

bigtedde replied:
I will give this a try - thank you :)

bigtedde commented on src/borg/testsuite/repo_new.py:
```diff
@@ -0,0 +1,352 @@
+    repository.delete(H(3))
+
+
+def repo_dump(repository, label=None):
```
Not used yet, but eventually will be used when more tests are added.

bigtedde commented on src/borg/testsuite/repo_new.py:
```diff
@@ -0,0 +1,352 @@
+    print()
+
+
+def test1(repository):
```
Renaming these is still on my todo list.

bigtedde commented on src/borg/testsuite/repo_new.py:
```diff
@@ -0,0 +1,352 @@
+    repository.reopen()
+    assert len(repository) == 3
+    # assert repository.check() is True - repository._active_txn is not false, causing check() to fail.
+    # my reopen method in repository isn't working as intended
```
`assert repository.check() is True` fails because `repository._active_txn` is not false, causing the check to fail. This is all due to my reopen method not working as intended. The original reopen method created a new repository object, which resets this variable to false. 

I went in loops for hours today trying to get this working correctly, going to sleep on it and try again in the morning.

RonnyPfannschmidt commented on conftest.py:
```diff
@@ -30,6 +32,13 @@ def clean_env(tmpdir_factory, monkeypatch):
 
+@pytest.fixture(autouse=True)
+def repository(tmpdir):
+    repository_location = os.path.join(str(tmpdir), "repository")
```
```suggestion
    repository_location = os.fspath(tmp_path/ "repository")
```

bigtedde commented on src/borg/testsuite/repo_new.py:
```diff
@@ -8,6 +10,27 @@
+    repository.__exit__(None, None, None)
+    reopened_repository = Repository(repository.path, exclusive=exclusive, create=False)
+    yield reopened_repository.__enter__()
+    repository.__exit__(None, None, None)
```
@RonnyPfannschmidt is this what you had in mind?

The only issue I run into now is that whenever a test uses this reopen, I get a warning that cleanup happened in `__del__`. Am I not exiting the repo correctly? Full warning below:

`Traceback (most recent call last):
    File "/Users/ted/Desktop/borg_workspace/borg/src/borg/repository.py", line 212, in __del__
      assert False, "cleanup happened in Repository.__del__"
  AssertionError: cleanup happened in Repository.__del__`
  
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))`
    
    @ThomasWaldmann @m3nu

bigtedde commented on src/borg/repository.py:
```diff
@@ -209,7 +209,7 @@ def __init__(
         if self.lock:
             self.close()
-            assert False, "cleanup happened in Repository.__del__"
+            # assert False, "cleanup happened in Repository.__del__"
```
Just to mute this warning until I can find a way to resolve. Any test that uses the `reopen_repository` will have this warning.

bigtedde commented on src/borg/testsuite/repo_new.py:
```diff
@@ -1,34 +1,51 @@
+
+def set_exclusive(value):
+    global exclusive
+    exclusive = value
```
These `get` and `set` are kind of a hacky way of changing the `exclusive` value, which some tests need changed when reopening a repository. I am working to find a more elegant way of doing this, but for now it works.

bigtedde commented on src/borg/testsuite/repo_new.py:
```diff
@@ -369,3 +396,488 @@ def test_crash_before_write_index(repository, reopened_repository):
+    with auxiliary_repository as repository:
+        with open(os.path.join(repository.path, "hints.1"), "ab") as fd:
+            fd.write(b"123456789")
+        do_commit(repository)
```
This test gives a warning, posted below. If I comment out this test, then the warning going on to the next test, and so on. I believe this has to do with an issue with my auxiliary_repository.

```
src/borg/testsuite/repo_new.py::test_corrupted_hints
  /~/python3.11/site-packages/_pytest/unraisableexception.py:78: PytestUnraisableExceptionWarning: Exception ignored in: <function Repository.__del__ at 0x102140d60>
  
  Traceback (most recent call last):
    File "/Users/ted/Desktop/borg_workspace/borg/src/borg/repository.py", line 211, in __del__
      self.close()
    File "/Users/ted/Desktop/borg_workspace/borg/src/borg/repository.py", line 488, in close
      self.lock.release()
    File "/Users/ted/Desktop/borg_workspace/borg/src/borg/locking.py", line 431, in release
      self._roster.modify(EXCLUSIVE, REMOVE)
    File "/Users/ted/Desktop/borg_workspace/borg/src/borg/locking.py", line 329, in modify
      elements.remove(self.id)
  KeyError: ('theodores-macbook-air.local@169793140525769', 43199, 0)
  
    warnings.warn(pytest.PytestUnraisableExceptionWarning(msg))
```

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -15,11 +15,29 @@
+
+    # teardown
+    repository.__exit__(None, None, None)
+    # shutil.rmtree(str(tmpdir))
```
This line caused the windows tests to fail on last commit, need to figure out why. Commented it out for now.

ThomasWaldmann replied:
check the tmpdir fixture docs - i think tmpdir does cleanup itself anyway.

ThomasWaldmann replied:
btw, guess some try/finally is missing here.

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -15,11 +15,29 @@
+
+@contextmanager
+def reopen(repository, exclusive=True):
+    try:
+        repository.__exit__(None, None, None)
+        repo = Repository(repository.path, exclusive=exclusive, create=False)
+        yield repo.__enter__()
+    finally:
+        repository.__exit__(None, None, None)
```
try try/finally does not look quite right.

guess it should be:
```
r = repo.__enter__()
try:
    yield r
finally:
    repo.__exit__(...)
```

ThomasWaldmann replied:
also, i think the repository fixture would not work correctly with the reopen CM:
- the fixture would enter the context1 before the test function starts
- reopen would exit context1, enter context2, exit context2
- the fixture would exit context1 after the test function finished

So there is one exit too much and also the stuff is not supposed to be nested into each other - we are dealing with the same repository for both contexts.

ThomasWaldmann replied:
after thinking a bit about it: if we want to have something like reopen, i don't think we can have an **"entered repository context" as a fixture**.

guess we could have **"repository" as a fixture** though. this still removes some code duplication, could clean up afterwards, etc. and would avoid the issues due to bad nesting. there could be also **"remote_repository"** as a fixture.

m3nu replied:
Yes, that's what I was thinking as well, when discussing this with Ted. Though I don't know the code as well, admittedly. So not sure it's possible speed-wise. But logically it would be best.

bigtedde replied:
> So there is one exit too much and also the stuff is not supposed to be nested into each other

This is a really good point, and something I had overlooked. This may be the source of some of the warnings I was encountering.

> after thinking a bit about it: if we want to have something like reopen, i don't think we can have an "entered repository context" as a fixture.

Are you suggesting to remove the `reopen` CM in place of something different (like just a function that reopens and returns?) to avoid poor nesting, or remove `reopen` entirely? 

> guess we could have "repository" as a fixture though. this still removes some code duplication, could clean up afterwards, etc. and would avoid the issues due to bad nesting

To be clear, are you suggesting to have `repository` do the initial setup, yield the repo, at end do final teardown, meanwhile entering the repo within the test instead of the in the repository fixture?

ThomasWaldmann replied:
the CM would be entered/exited within the test function (not within the fixture code).

the whole point of `reopen()` was to get a really completely fresh Repository instance, so we do not rely on previous state except the files that were stored on disk.

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -716,339 +767,373 @@ def _subtly_corrupted_hints_setup(self):
+    assert remote_repository.ssh_cmd(Location("ssh://example.com/foo")) == ["ssh", "-i", "foo", "example.com"]
+
+
+""" REMOTE REPOSITORY CHECK TESTS"""
```
guess one rather wants to use a comment for this.

one docstring on module level (usually at top of the module code) is rather for documenting what the module does.

ThomasWaldmann reviewed (changes requested):
guess that fixture/contextmanager combination is problematic.

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -15,11 +13,21 @@
+def repository(tmpdir):
+    repository_location = os.path.join(str(tmpdir), "repository")
+    repository = Repository(repository_location, exclusive=True, create=True)
+    yield repository
```
If it wasn't for needing to reopen the repository during a few tests, I would have done something like:
```
with Repository(repository_location, exclusive=True, create=True) as repository
     yield repository
```
since this CM would automatically enter and exit the repository context. But I can't figure out how to get `reopen()` to work with this.

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -15,11 +13,21 @@
+    yield repository
+
+
+def reopen(repository, exclusive=True):
+    repository.__exit__(None, None, None)
+    repository = Repository(repository.path, exclusive=exclusive, create=False)
+    repository.__enter__()
+    return repository
```
I can make that happen, thank you!

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -15,11 +14,24 @@
+    yield repository
+
+
+@contextmanager
+def reopen(repository, exclusive=True):
+    repo = Repository(repository.path, exclusive=exclusive, create=False)
+    repo.__enter__()
+    try:
+        yield repo
+    finally:
+        repo.__exit__(None, None, None)
```
hmm, guess you have 1 context manager too much here:
- the Repository instance you have in `repo` is already a CM (== it has `__enter__` and `__exit__` methods)
- you are creating another CM here, wrapping the other one and doing nothing else

guess that would also work:
```
def reopen(repository, exclusive=True):
    return Repository(repository.path, exclusive=exclusive, create=False)


def test(repository):
    with repository:
        repository.do_stuff()
    with reopen(repository) as repository:
        repository.check_stuff()
```

bigtedde replied:
@ThomasWaldmann Thank you for your patience as I work to understand CM's better. I see now why the repo's `__enter__` and `__exit__` weren't getting called automatically during the `reopen` the same way they were with the original repository fixture, because I had wrapped them in an additional CM. Thank you for clarifying that.

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -716,339 +793,370 @@ def _subtly_corrupted_hints_setup(self):
+@pytest.fixture()
+def remote_repository(tmpdir):
+    repository_location = Location("ssh://__testsuite__" + os.path.join(str(tmpdir), "repository"))
+    with RemoteRepository(repository_location, exclusive=True, create=True) as remote_repository:
+        yield remote_repository
```
guess this needs an update.

for the `repository` fixture, we have learned that we **must not** enter the CM within the fixture, but just return the CM and do the entering in the test function (using `with` there).

bigtedde replied:
Makes sense! I had left this as it was because we did not have any 'reopen' calls in the remote repo tests, so figured it would be fine. However, I now see what you mean, and will make this change 👍

RonnyPfannschmidt commented:
this code with all those __enter/__exit__ calls is absolutely terrifying and not at all what i meant

i was thinking closer to something like
```python


class SingularRepositoryMaker:
   args: ...
   
   def __enter__(self):
      if self.current_repo is not None:
         raise RuntimeError(...)
      self.current_repo = self._exitstack.enter_context(Repository(**self.args))
     
      return current_repo
      
    def __exit__(self,...):
        self.context_stack.undo() # perhaps apass over args here
        self.current_repo = None
      

```

also if reentering a once disposed repository is not safe, lets make sure a exception is thrown if __enter__ is done a second time on repository

bigtedde commented:
> self.context_stack.undo()  # perhaps pass over args here

Is this where the repo's exit would get called? I'm having trouble seeing how to undo the actions here.

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -716,339 +793,370 @@ def _subtly_corrupted_hints_setup(self):
+
+@pytest.fixture()
+def remote_repository(tmpdir):
+    repository_location = Location("ssh://__testsuite__" + os.path.join(str(tmpdir), "repository"))
```
I believe they are filtered out:

`pytest -n4 --benchmark-skip -vv -rs -k "not remote"`

Previously they were in a class with the word 'remote' in the name, so I believe they were skipped. I will try changing the test names to include 'remote' and see that will skip them going forward.

m3nu replied:
Another [option](https://docs.pytest.org/en/7.3.x/how-to/skipping.html) to skip some tests on Windows

`@pytest.mark.skipif(sys.platform == "win32", reason="does not run on windows")`

ThomasWaldmann replied:
In `borg.platform_flags`, we have "ready to use" `is_win32` and similar flags.

See line 1065.

bigtedde replied:
Thanks, I will include this in the remote_repository fixture so all tests that use it will be skipped on windows,

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -15,11 +15,39 @@
+    def __exit__(self, exc_type, exc_value, traceback):
+        # Decide what needs to be done here.
+        # self.context_stack.undo()  # perhaps pass over args here
+        self.current_repo = None
```
Per @RonnyPfannschmidt's suggestion, however I'm not sure how to incorporate this into what we have yet, but wanted to leave in for ideas/suggestions. Currently this class is not used.

m3nu replied:
To pick up on @RonnyPfannschmidt's comment regarding this code: I think he's right in that the `__enter/__exit` stuff isn't very clear. But as Ted says, this code isn't used right now.

If that's the case, the Pytest `repository` fixture and the context manager seem to do the same job and we can get rid of one of them. In detail, the fixture an already run some code before and after a test and `__enter/__exit` stuff can be run there. If we have the case where one test needs multiple repos, maybe we should split those up?

bigtedde commented:
@ThomasWaldmann Thanks again for all your support and feedback. I fixed the 'reopen' CM, remote repo CM, and renamed the remote repo tests so that they will be skipped on Windows. I feel like things are progressing and in a good place now. Next, I am going to work on incorporating Ronnys ideas 👍

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -15,11 +15,39 @@
+
+    def __enter__(self):
+        if self.current_repo is not None:
+            raise RuntimeError("Cannot re-enter a disposed repository.")
```
"can not support nested repository contexts" maybe?

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -716,339 +809,365 @@ def _subtly_corrupted_hints_setup(self):
+        make_auxiliary(repository)
+    _subtly_corrupted_hints_setup(repository)
+
+    with repository:
```
should that be a reopen?

ThomasWaldmann reviewed (commented):
did only a quick check, but structurally this now looks good.

did you check how many tests total it executes before and after your changes?

bigtedde commented:
> did only a quick check, but structurally this now looks good.
> 
> did you check how many tests total it executes before and after your changes?

Original test file ran 81 tests, mine was only 66. I realized today that the original file was running the first 15 or so tests (from `class RepositoryTestCase`) on both the regular Repository as well as RemoteRepository. I am working now to parameterize these tests to run on both types of repos, and exclude the remote repo tests from windows.

bigtedde commented:
The new repo tests now cover all 84 of the original repo tests. I found that several tests were being ran on both Repository as well as Remote Repository (something I had been overlooking) and so I parameterized these tests to run both. 

With feedback from Thomas and Manu, I included a skip in the remote repo fixture for windows.

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -94,8 +91,10 @@ def repo_dump(repository, label=None):
 
-def test_basic_operations(repository):
-    with repository:
+@pytest.mark.parametrize("fixture", ["repository", "remote_repository"])
```
For the tests that were previously ran with both a local and remote repo, I decided to use parameterization to accomplish the same thing.

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -655,7 +698,7 @@ def make_auxiliary(repository):
 
 def do_commit(repository):
-    with repository:
+    with reopen(repository) as repository:
```
Since do_commit was using a CM, I think it is best to reopen the repo? In the tests that call this, I ensure that it is not called within other CM's.

ThomasWaldmann replied:
Not sure I understand the question.

The misc. tests calling `do_commit` get a fresh `repository` each.

Also as the repo CM is opening/closing the repo, i think the commit has to happen while the repo context is entered (while the repo is opened).

bigtedde replied:
Each of the do_commit calls in the tests happen after the original repo CM has exited though, so wouldn't we be reopening the repo in these cases?

ThomasWaldmann replied:
yes, indeed.

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -1171,3 +1221,18 @@ def test_remote_borg_cmd(remote_repository):
+
+def test_remote_crash_before_compact():
+    # skip this test, we can't mock-patch a Repository class in another process!
+    pass
```
I guess we could just as soon delete these last 3 tests? Their use as placeholders I think was a unittest thing to prevent execution of other tests with remote repo.

ThomasWaldmann replied:
The unittest-style test class for the remote repository tests had inherited most tests from the local repository test class, but some tests overwritten to not execute them because they can't work (see comments).

The comments might be worth keeping at the new place, where you do **not** have the remote_repo fixture now, to explain "why".

bigtedde replied:
Good point :)

ThomasWaldmann replied:
The comments in such "skipped" / empty tests still need to get moved to the right place.

The right place now is above or in the test function where you chose not to parametrize for both local and remote repos to get tested, but just tested for a local repo. So you can give a comment **there** to tell **why you didn't test with the remote repo**.

bigtedde replied:
That makes a lot of sense, thank you. I'll get that changed now.

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -12,42 +11,40 @@
+    return RemoteRepository(repo_location, exclusive=True, create=True)
+
+
+def reopen(repository, exclusive: Optional[bool] = True, create=False):
```
Maybe we could put some sort of check here to ensure the repository we are reopening has already been closed?

ThomasWaldmann replied:
Yes. Guess that was also one part of what ronny suggested.

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -94,8 +91,10 @@ def repo_dump(repository, label=None):
-    with repository:
+@pytest.mark.parametrize("fixture", ["repository", "remote_repository"])
+def test_basic_operations(fixture, request):
+    repo = request.getfixturevalue(fixture)
+    with repo as repository:
```
Interesting. Not sure if that is the best way to accomplish that, maybe @RonnyPfannschmidt knows better.

RonnyPfannschmidt replied:
the lazy_fixtures plugin has a way to refer to other fixtures as parameters safely,
i generally recommend against using getfixturevalue and request inside of a test function

ThomasWaldmann replied:
doesn't look maintained, so we'll do it without it for now.

RonnyPfannschmidt commented on src/borg/testsuite/repository.py:
```diff
@@ -15,11 +15,39 @@
+    # For now, this class is not used. Just a placeholder until I am sure if/how to incorporate it into project.
+    def __init__(self, exclusive=True, create=True):
+        self.current_repo = None
+        self.path = os.path.join(tempfile.mkdtemp(), "repository")
```
when using pytest tmp_path shoudl be used - mkdtemp would leak its a antipattern to use it under ptest

RonnyPfannschmidt commented on src/borg/testsuite/repository.py:
```diff
@@ -12,42 +11,40 @@
-def reopen(repository, exclusive=True, create=False):
-    return Repository(repository.path, exclusive=exclusive, create=create)
+@pytest.fixture()
+def remote_repository(tmpdir):
```
use tmp_path instead of tmpdir

RonnyPfannschmidt reviewed (commented):
im under the impression, there should be somethinh like a testrepositorycontext, that on entering would make new repository instances

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -19,24 +19,28 @@
         return Repository(repository.path, exclusive=exclusive, create=create)
     elif isinstance(repository, RemoteRepository):
+        if repository.p or repository.sock is not None:  # The remote repo should be closed before reopening
+            raise RuntimeError("Cannot support nested repository contexts.")
```
In Repository, `close()` sets `io` and `lock` to `None`, so if either of these aren't `None` then the repo should not be reopened.

In RemoteRepository, `close` sets `p` and `sock` to None, so same deal here.

ThomasWaldmann replied:
You need to check `.io` and `.p` using `is not None`.

bigtedde replied:
On it 👍 Thank you.

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -544,17 +547,17 @@ def test_shadowed_entries_are_preserved(repository):
         repository.delete(H(1))
         repository.put(H(2), fchunk(b"1"))
-        delete_segment = get_latest_segment()
+        del_segment = get_latest_segment()
```
delete_segment shadowed a helper function name from outer scope. Renamed for clarity.

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -91,10 +95,9 @@ def repo_dump(repository, label=None):
-def test_basic_operations(fixture, request):
-    repo = request.getfixturevalue(fixture)
-    with repo as repository:
+@pytest.mark.parametrize("fixture", [pytest.lazy_fixture("repository"), pytest.lazy_fixture("remote_repository")])
```
This works, however when discussing with @m3nu we were skeptical about including this as it seems like a small project that could be compromised: https://github.com/TvoroG/pytest-lazy-fixture

m3nu replied:
Or abandoned and unmaintained, which sadly happens too often for useful packages.

ThomasWaldmann replied:
That lazy-fixture project doesn't look like regularly maintained. Last release 3.5y old.

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -45,1010 +74,1164 @@ def pdchunk(chunk):
+        assert repository.storage_quota_use == len(ch1) + 41 + 8  # now we have compacted.
+
 
+def make_auxiliary(repository):
+    repository.put(H(0), fchunk(b"foo"))
+    repository.commit(compact=False)
+    repository.close()
```
Or maybe it also makes sense to enter/leave the repository context inside `make_auxiliary`.

ThomasWaldmann reviewed (commented):
please review these "corruption tests" concerning the correct opening, reopening, closing of the repo, where a commit/rollback happens, that stuff that does not need to be within the repo context is happening outside the repo context.

bigtedde commented:
> please review these "corruption tests" concerning the correct opening, reopening, closing of the repo, where a commit/rollback happens, that stuff that does not need to be within the repo context is happening outside the repo context.

I am reviewing this now, and will have this taken care of by end of day 👍

bigtedde commented:
Hmmm. Windows test didn't pass -

`borg.testsuite.archiver.extract_cmd.ArchiverTestCase testMethod=test_sparse_file`

It provides assertion error that looks similar to one encountered the other day (something to do with a Location object, iirc)

```
with open(filename, "rb") as fd:
                # check if file contents are as expected
>               self.assert_equal(fd.read(hole_size), b"\0" * hole_size)
E               AssertionError: b'\x0[8388600 chars]0\x00|\xbe \xe0\xc1\xfcg\xbf\xf1bGF\x931\x8c\x[159271271 chars]\x00' != b'\x0[8388600 chars]0\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\[159383507 chars]\x00'
```

I looked through the test_spare_file test and couldn't find anything that would have broken from my changes 🤔. Will keep my eyes peeled.

ThomasWaldmann commented:
@bigtedde 

> Hmmm. Windows test didn't pass -
> `borg.testsuite.archiver.extract_cmd.ArchiverTestCase testMethod=test_sparse_file`

see #7616.

bigtedde commented:
Made some minor clean up changes today, and after a new commit they got past the spurious Windows error. 

@ThomasWaldmann I am feeling really good about where we are with this test conversion. Do you have any areas of concern I could address before marking this PR ready for review?

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -12,14 +11,58 @@
+    if is_win32:
+        pytest.skip("On Windows, we skip the remote tests.")
+    repo_location = Location("ssh://__testsuite__" + os.fspath(tmp_path / "repository"))
+    return RemoteRepository(repo_location, exclusive=True, create=True)
```
>I just noticed that you return here - guess you should rather yield

Whoops, not sure how I let that slip. Thanks for catching.

bigtedde commented on src/borg/testsuite/repository.py:
```diff
@@ -328,11 +327,11 @@ def test_flags_persistence(repo_fixtures, request):
 
 def _assert_sparse(repository):
-    # The superseded 123456... PUT
+    # the superseded 123456... PUT
```
Most comments started with lowercase, so for consistency I made them all uniform lowercase.

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -12,14 +11,57 @@
+
+    # reopening a closed remote repo
+    elif isinstance(repository, RemoteRepository):
+        if repository.p is not None and repository.sock is not None:  # this ensures remote repo is closed
```
guess this needs to be `or`.

bigtedde replied:
Definitely should be `or`, I'll fix that.

ThomasWaldmann replied:
same for the comment here.

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -12,14 +11,57 @@
+def reopen(repository, exclusive: Optional[bool] = True, create=False):
+    # reopening a closed local repo
+    if isinstance(repository, Repository):
+        if repository.io is not None and repository.lock is not None:  # this ensures the repo is closed
```
`or`?

ThomasWaldmann replied:
also, guess it is better to remove that comment. it says the opposite of what the "if" is checking for.
also, the RuntimeError string already says what's the issue here.

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -33,26 +33,24 @@ def remote_repository(tmp_path):
 
 def pytest_generate_tests(metafunc):
-    # some tests are run on both local and remote repos
+    # parameterizes tests that run on both local and remote repos
```
parametrize (see method name)?

bigtedde replied:
I worded it that way because it replaces parameterization in the tests that use it, but maybe "generates" would be more accurate here?

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -12,14 +11,55 @@
+
+
+def reopen(repository, exclusive: Optional[bool] = True, create=False):
+    if isinstance(repository, Repository):
+        if repository.io is not None or repository.lock is not None:
+            raise RuntimeError("Repo must be closed before a reopen. Cannot support nested repository contexts.")
+        return Repository(repository.path, exclusive=exclusive, create=create)
+
+    elif isinstance(repository, RemoteRepository):
+        if repository.p is not None or repository.sock is not None:
+            raise RuntimeError("Remote repo must be closed before a reopen. Cannot support nested repository contexts.")
+        return RemoteRepository(repository.location, exclusive=exclusive, create=create)
```
i'ld change 52 to "if ..." (the "else" is given since there is no other way to get here).

also: add a `raise TypeError(...)` after line 55.

**alternatively**: if ... / elif ... / else: raise

bigtedde replied:
Gotcha - I'm on it!

ThomasWaldmann commented on src/borg/testsuite/repository.py:
```diff
@@ -12,14 +11,55 @@
+
+
+def get_path(repository):
+    if isinstance(repository, Repository):
+        return repository.path
+    elif isinstance(repository, RemoteRepository):
+        return repository.location.path
```
similar here

ThomasWaldmann reviewed (changes requested):
ok, last review, will merge after final fixes.

ThomasWaldmann commented:
squash all commits together and keep what looks valuable from all the commit comments?

ThomasWaldmann commented:
Thanks!
